firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -198,4 +215,48 @@ void shutdown() {
       Thread.currentThread().interrupt();
     }
   }
+
+  /**
+   * Sets the connectivity attempt timer to track. Existing timers must complete or be cancelled
+   * before a new timer can be set.
+   */
+  void setConnectivityAttemptTimer(DelayedTask connectivityAttemptTimer) {
+    // If an existing timer is already running, we want to continue using that one.
+    if (this.connectivityAttemptTimer != null) {
+      connectivityAttemptTimer.cancel();
+    } else {
+      try {
+        ManagedChannel channel = this.channelTask.getResult();
+        if (channel != null) {
+          this.connectivityAttemptTimer = connectivityAttemptTimer;
+          listenToConnectivityState(channel);
+        }
+      } catch (IllegalStateException e) {
+        // When the ManagedChannel is first initialized, channelTask could still be running.
+        connectivityAttemptTimer.cancel();
+      }
+    }
+  }
+
+  /**
+   * Cancels the connectivityStateTimer if the new state indicates grpc is online. Otherwise, we
+   * reset the listener for the next state change.
+   */
+  private void listenToConnectivityState(ManagedChannel channel) {
+    ConnectivityState newState = channel.getState(false);
+    // Check that the new state is online, then cancel timer.
+    if (newState == ConnectivityState.READY) {
+      clearConnectivityTimer();
+    } else {
+      channel.notifyWhenStateChanged(newState, () -> listenToConnectivityState(channel));
+    }
+  }
+
+  /** Clears the connectivity timer if it exists. */
+  void clearConnectivityTimer() {
```

Comment:
This is named inconsistently with setConnectivityAttemptTimer, though they seem related. Also, they could likely be grouped together.

Reply:
Done.